File: NeoForge/src/main/java/team/zeds/ancientmagic/forge/api/HandleStack.java
package team.zeds.ancientmagic.forge.api;

import net.minecraft.core.NonNullList;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.world.Container;
import net.minecraft.world.SimpleContainer;
import net.minecraft.world.item.ItemStack;
import net.minecraftforge.items.ItemStackHandler;
import org.apache.commons.lang3.ArrayUtils;
import org.jetbrains.annotations.NotNull;
import team.zeds.ancientmagic.common.api.helper.IHandleStack;

import java.util.HashMap;
import java.util.Map;
import java.util.function.BiFunction;
import java.util.function.Consumer;
import java.util.function.Function;

public class HandleStack extends ItemStackHandler implements IHandleStack<HandleStack> {
    private final Runnable onContentsChanged;
    private final Map<Integer, Integer> slotSizeMap;
    private BiFunction<Integer, ItemStack, Boolean> canInsert = null;
    private Function<Integer, Boolean> canExtract = null;
    private int maxStackSize = 64;
    private int[] outputSlots = null;

    protected HandleStack(int size, Runnable onContentsChanged) {
        super(size);
        this.onContentsChanged = onContentsChanged;
        this.slotSizeMap = new HashMap<>();
    }

    @Override
    public @NotNull ItemStack insertItem(int slot, @NotNull ItemStack stack, boolean simulate) {
        return this.insertItem(slot, stack, simulate, false);
    }

    @NotNull
    @Override
    public ItemStack insertItem(int slot, @NotNull ItemStack stack, boolean simulate, boolean container) {
        if (!container && this.outputSlots != null && ArrayUtils.contains(this.outputSlots, slot))
            return stack;
        return super.insertItem(slot, stack, simulate);
    }

    @Override
    public ItemStack extractItem(int slot, int amount, boolean simulate) {
        return this.extractItem(slot, amount, simulate, false);
    }

    public ItemStack extractItem(int slot, int amount, boolean simulate, boolean container) {
        if (!container) {
            if (this.canExtract != null && !this.canExtract.apply(slot))
                return ItemStack.EMPTY;

            if (this.outputSlots != null && !ArrayUtils.contains(this.outputSlots, slot))
                return ItemStack.EMPTY;
        }

        return super.extractItem(slot, amount, simulate);
    }

    @Override
    public int getSlotLimit(int slot) {
        return this.slotSizeMap.containsKey(slot) ? this.slotSizeMap.get(slot) : this.maxStackSize;
    }

    @Override
    public boolean isItemValid(int slot, @NotNull ItemStack stack) {
        return this.canInsert == null || this.canInsert.apply(slot, stack);
    }

    @Override
    protected void onContentsChanged(int slot) {
        if (this.onContentsChanged != null)
            this.onContentsChanged.run();
    }

    @NotNull
    @Override
    public NonNullList<ItemStack> getStacks() {
        return this.stacks;
    }

    @NotNull
    @Override
    public int[] getOutputSlots() {
        return this.outputSlots;
    }

    @Override
    public void setDefaultSlotLimit(int size) {
        this.maxStackSize = size;
    }

    @Override
    public void addSlotLimit(int slot, int size) {
        this.slotSizeMap.put(slot, size);
    }

    @Override
    public void setCanInsert(@NotNull BiFunction<Integer, ItemStack, Boolean> validator) {
        this.canInsert = validator;
    }

    @Override
    public void setCanExtract(@NotNull Function<Integer, Boolean> canExtract) {
        this.canExtract = canExtract;
    }

    @Override
    public void setOutputSlots(@NotNull int... slots) {
        this.outputSlots = slots;
    }

    @NotNull
    @Override
    public Container toContainer() {
        return new SimpleContainer(this.stacks.toArray(new ItemStack[0]));
    }

    @NotNull
    @Override
    public HandleStack copy() {
        final var newInv = new HandleStack(this.getSlots(), this.onContentsChanged);

        newInv.setDefaultSlotLimit(this.maxStackSize);
        newInv.setCanInsert(this.canInsert);
        newInv.setCanExtract(this.canExtract);
        newInv.setOutputSlots(this.outputSlots);

        this.slotSizeMap.forEach(newInv::addSlotLimit);

        for (int i = 0; i < this.getSlots(); i++) {
            var stack = this.getStackInSlot(i);

            newInv.setStackInSlot(i, stack.copy());
        }

        return newInv;
    }

    @Override
    public void deserializeTag(@NotNull CompoundTag tag) {
        this.deserializeNBT(tag);
    }

    @NotNull
    @Override
    public CompoundTag serializeTag() {
        return this.serializeNBT();
    }

    @NotNull
    @Override
    public ItemStack getStackInSlotHandler(int slot) {
        return this.getStackInSlot(slot);
    }

    @NotNull
    @Override
    public HandleStack nsCreate(int size) {
        return HandleStack.create(size);
    }

    @Override
    public void setSizeHandler(int i) {
        this.setSize(i);
    }

    public static HandleStack create(int size) {
        return create(size, builder -> {});
    }

    public static HandleStack create(int size, Runnable contentChange) {
        return create(size, contentChange, builder -> {});
    }

    public static HandleStack create(int size, Consumer<HandleStack> builder) {
        return create(size, null, builder);
    }

    public static HandleStack create(int size, Runnable contentChange, Consumer<HandleStack> builder) {
        var handler = new HandleStack(size, contentChange);
        builder.accept(handler);
        return handler;
    }
}

File: src/main/java/team/zeds/ancientmagic/api/handler/HandleStack.kt
package team.zeds.ancientmagic.api.handler

import net.minecraft.core.NonNullList
import net.minecraft.world.Container
import net.minecraft.world.SimpleContainer
import net.minecraft.world.item.ItemStack
import net.minecraftforge.items.ItemStackHandler
import org.apache.commons.lang3.ArrayUtils
import java.util.function.BiFunction
import java.util.function.Consumer
import java.util.function.Function

open class HandleStack protected constructor(size: Int, @JvmField val contentChange: Runnable): ItemStackHandler(size) {
    @JvmField val slotSizeMap: MutableMap<Int, Int>
    @JvmField var canInsert: BiFunction<Int, ItemStack, Boolean>? = null
    @JvmField var canExtract: Function<Int, Boolean>? = null
    @JvmField var maxStackSize = 64
    @JvmField var outputSlots: IntArray? = null

    init {
        slotSizeMap = HashMap()
    }

    override fun insertItem(slot: Int, stack: ItemStack, simulate: Boolean): ItemStack =
        this.insert(slot, stack, simulate, false)


    open fun insert(slot: Int, stack: ItemStack, simulate: Boolean, container: Boolean): ItemStack {
        if (!container && this.outputSlots != null && ArrayUtils.contains(this.outputSlots, slot)) return stack
        return super.insertItem(slot, stack, simulate)
    }

    override fun extractItem(slot: Int, amount: Int, simulate: Boolean): ItemStack =
        extract(slot, amount, simulate, false)

    open fun extract(slot: Int, amount: Int, simulate: Boolean, container: Boolean): ItemStack {
        if (!container) {
            if (this.canExtract != null && !this.canExtract!!.apply(slot)) return ItemStack.EMPTY
            if (this.outputSlots != null && !ArrayUtils.contains(this.outputSlots, slot)) return ItemStack.EMPTY
        }
        return super.extractItem(slot, amount, simulate)
    }

    override fun getSlotLimit(slot: Int): Int = if (this.slotSizeMap.containsKey(slot)) this.slotSizeMap[slot]!! else this.maxStackSize

    override fun isItemValid(slot: Int, stack: ItemStack): Boolean {
        return canInsert == null || canInsert!!.apply(slot, stack)
    }

    override fun onContentsChanged(slot: Int) {
        this.contentChange.run()
    }

    open fun getStacks(): NonNullList<ItemStack?>? {
        return stacks
    }

    open fun getOutputSlots(): IntArray? {
        return outputSlots
    }

    open fun setDefaultSlotLimit(size: Int) {
        maxStackSize = size
    }

    open fun addSlotLimit(slot: Int, size: Int) {
        slotSizeMap.put(slot, size)
    }

    open fun setCanInsert(validator: BiFunction<Int, ItemStack, Boolean>) {
        canInsert = validator
    }

    open fun setCanExtract(canExtract: Function<Int, Boolean>) {
        this.canExtract = canExtract
    }

    open fun setOutputSlots(vararg slots: Int) {
        outputSlots = slots
    }

    open fun toContainer(): Container {
        return SimpleContainer(*stacks.toTypedArray<ItemStack>())
    }

    open fun copy(): HandleStack {
        val newInventory = HandleStack(this.slots, this.contentChange)

        newInventory.setDefaultSlotLimit(maxStackSize)
        newInventory.setCanInsert(canInsert!!)
        newInventory.setCanExtract(canExtract!!)
        newInventory.setOutputSlots(*outputSlots!!)

        slotSizeMap.forEach(newInventory::addSlotLimit)

        for (i in 0 until this.slots) {
            val stack = getStackInSlot(i)
            newInventory.setStackInSlot(i, stack.copy())
        }

        return newInventory
    }

    companion object {
        @JvmStatic fun create(size: Int): HandleStack = create(size) { _ -> }
        @JvmStatic fun create(size: Int, contentChange: Runnable): HandleStack = create(size, contentChange) { _ -> }
        @JvmStatic fun create(size: Int, builder: Consumer<HandleStack>): HandleStack = create(size, null, builder)

        @JvmStatic
        fun create(size: Int, contentChange: Runnable?, builder: Consumer<HandleStack>): HandleStack {
            val handler = HandleStack(size, contentChange!!)
            builder.accept(handler)
            return handler
        }
    }
}